### PR TITLE
Enhanced volume validation messages

### PR DESCRIPTION
**Pull request by unterstein**
If the validation of a volume failed, the validation should return a readable error message.

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/core/externalvolume/impl/providers/DVDIProvider.scala [hunk 1]
```diff
@@ -182,7 +182,8 @@ private[impl] object DVDIProviderValidations extends ExternalVolumeValidations {
         volume.external.size is isTrue("must be undefined for Docker containers")(_.isEmpty)
         volume.containerPath is notOneOf(DotPaths: _*)
         // TODO(jdef) change this once docker containerizer supports relative containerPaths
-        volume.containerPath should matchRegexFully(AbsolutePathPattern)
+        volume.containerPath should
+          matchRegexFullyWithMessage(AbsolutePathPattern, "containerPath must not starts with \"/\"")
```
Comment (jdef) [inline src/main/scala/mesosphere/marathon/core/externalvolume/impl/providers/DVDIProvider.scala]: s/starts/start/

Comment (jdef) [inline src/main/scala/mesosphere/marathon/core/externalvolume/impl/providers/DVDIProvider.scala]: is `containerPath` really needed in these messages? doesn't accord add that in automatically when it parses the object hiearchy (object == the subject of validation)?

File: src/test/scala/mesosphere/mesos/PersistentVolumeValidationTest.scala [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    When("The volume is created and validation failed")
+    volume should not be null
+    volume.containerPath should be (path)
+    val validation = Volume.validVolume(volume)
+    validation.isSuccess should be (false)
+
+    Then("A validation exists with a readable error message")
+    validation match {
+      case Failure(violations) =>
+        violations should contain (RuleViolation("/path", "containerPath must not contain \"/\"", Some("containerPath")))
```
Comment (jdef) [inline src/test/scala/mesosphere/mesos/PersistentVolumeValidationTest.scala]: as per the above, if `containerPath` weren't part of the message there wouldn't be the redundant appearance of `containerPath` on this line

File: src/main/scala/mesosphere/marathon/api/v2/Validation.scala [hunk 3]
```diff
@@ -280,4 +283,17 @@ object Validation {
       validator(t)
     }
   }
+
+  /**
+    * A validator that succeeds only if the provided string matches the specified pattern.
+    *
+    * Initially took from com.wix.accord.combinators.StringCombinators to enhance validationMessage feature
+    */
+  class MatchesRegexWithMessage(pattern: Pattern, partialMatchAllowed: Boolean = true, validationMessage: String)
```
Comment (jdef) [inline src/main/scala/mesosphere/marathon/api/v2/Validation.scala]: how about a unit test that exercises partial vs. non-partial matching?

**Comment by jdef** [comment]
looking good so far, left some comments

File: src/main/scala/mesosphere/marathon/core/externalvolume/impl/providers/DVDIProvider.scala (hunk 1 above)
Comment (unterstein) [inline src/main/scala/mesosphere/marathon/core/externalvolume/impl/providers/DVDIProvider.scala]: Maybe something like `value must not start with "/"` ?

File: src/main/scala/mesosphere/marathon/api/v2/Validation.scala [hunk 4]
```diff
@@ -280,4 +283,17 @@ object Validation {
...
+    extends NullSafeValidator[String](
+      v => if (partialMatchAllowed) pattern.matcher(v).find() else pattern.matcher(v).matches(),
+      v => v -> validationMessage)
+
+  def matchRegexFullyWithMessage(regex: Regex, message: String): Validator[String] =
```
Comment (meichstedt) [inline src/main/scala/mesosphere/marathon/api/v2/Validation.scala]: I think you can boil this down to
```
  def regexValidator(regex: Regex, failureMessage: String) = new NullSafeValidator[String](
    test = _.matches(regex.regex),
    failure = _ -> failureMessage
  )
```
withouth the need for `MatchesRegexWithMessage `.

**Comment by meichstedt** [comment]
LGTM too – please remove the extra unneeded complexity in `Validation.scala` and this is ready merge.

**Comment by unterstein** [comment]
done

**Comment by meichstedt** [comment]
👍

**Pull request merged**